---
title: "HotSpot_analysis"
output: html_document
date: "2024-03-16"
editor_options: 
  chunk_output_type: console
---

**Start interactive session on Midway3**
```{bash}
ssh -Y zemmour@midway3.rcc.uchicago.edu
sinfo -s
screen -S david
sinteractive --exclusive --time=12:00:00 --account=pi-zemmour --partition=jfkfloor2 --qos=jfkfloor2 --nodes=1 
sinteractive --time=12:00:00 --account=pi-zemmour --partition=caslake --nodes=1 --mem=128GB
sinteractive --time=3:00:00 --account=pi-zemmour --partition=gpu --gres=gpu:1 --mem=152GB
cd $LABHOME

module load python/anaconda-2022.05 
module load hdf5/1.12.0 #for BPCells but also others
module load openblas/0.3.13

#sinteractive --time=1:00:00 -p gpu --gres=gpu:1 --account=pi-zemmour 

#source activate /project/jfkfloor2/zemmourlab/david/envs/scvi
source activate /project/zemmour/david/envs/scvi_20240315
module load openblas/0.3.13 #load again or error

cd /project/jfkfloor2/zemmourlab/david/immgent/analysis/integration/IGT1_56/Treg
#cd ~/google_drive/ImmgenT/analysis/data_integration/IGT1_56/Treg

#R
#options(max.print=1000)
#options(Seurat.object.assay.version = 'v5')
```

**Initialize in R: load libraries and custom functions**

```{r}
setwd("~/google_drive/ImmgenT/jamboree2/")
setwd("~/google_drive/ImmgenT/analysis/data_integration/IGT1_56")
setwd("/project/jfkfloor2/zemmourlab/david/immgent/analysis/integration/IGT1_56/")

libs = c("Seurat","BPCells", "ggplot2", "dplyr", "reshape2", "ggrastr", "RColorBrewer", "pals", "scales", "pheatmap") 
#libs = c("matrixStats", "gplots","ggplot2", "reshape2", "scales", "gridExtra", "dplyr", "RColorBrewer", "grid", "Rtsne", "limma",   "RColorBrewer", "pheatmap", "Seurat",  "ggrastr", "ggbeeswarm") #pwr, preprocessCore, "readxl", vegan", "genefilter" "scde" "cellrangerRkit", "Rsamtools", "GenomicRanges", "GenomicAlignments", "VGAM", "WGCNA",rafalib, ‘UsingR’,MAST "ggrastr",
sapply(libs, function(x) suppressMessages(suppressWarnings(library(x, character.only = TRUE, quietly = T, warn.conflicts  = F))))

options(Seurat.object.assay.version = 'v5')

ColorRamp = rev(colorRampPalette(brewer.pal(n = 7,name = "RdYlBu"))(100))
ColorRamp = rev(colorRampPalette(c('red','white','blue'))(20))
ColorRamp = rev(rainbow(10, end = 4/6))
ColorRamp = viridis(100)

#mycol = c("black","blue","red","darkgreen", "orange","purple", "cyan", "greenyellow",    "salmon", "magenta","pink", "tan", "brown") # c("magenta", "red",  "darkgreen", "cyan", "blue", "blue", "blue", "black", "blue", "blue", "blue", "blue", "orange")

n = 70
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
mypal1 = unique(unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals))))
mypal1 = mypal1[-4]

library("pals")
parade = function(n) { return(Seurat::DiscretePalette(n, palette = "parade", shuffle = F)) }

length(glasbey())
length(polychrome())
mypal = c(glasbey(), polychrome(), mypal1)
names(mypal) = NULL
#pal.bands(glasbey(), polychrome(), parade(), main="Colormap suggestions")

#pdf("mypal.pdf", 10, 10)
#pie(rep(1,n), col=unique(unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))) )
#pie(rep(1,n), col=unique(unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))), labels =  unique(unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))))
#dev.off()

MyPlots = function (seurat_object = so, dim1 = so[["umap_unintegrated"]]@cell.embeddings[,1], dim2 = so[["umap_unintegrated"]]@cell.embeddings[,2], color_by = "spleen_standard", split_by1 = "IGT", split_by2 =  NULL, genes = c("Foxp3", "Il2ra"), cluster_key = "ClusterSCVI_Res", mypal = glasbey()) {

    so = seurat_object
    #so@meta.data[,"split_by"] = so@meta.data[,split_by]
    so@meta.data[,"color_by"] = factor(so@meta.data[,color_by])
    so@meta.data[,"split_by1"] = so@meta.data[,split_by1]
    so@meta.data[,"split_by2"] = so@meta.data[,split_by2]
    
    alpha = 0.5
    #sample_name_colors = color_palette[1:length(unique(so@meta.data[,color_by]))]
    #names(sample_name_colors) = levels(so$sample_name)
    #sample_name_colors2 = sample_name_colors
    #sample_name_colors2[grepl("WT", names(sample_name_colors))] = "grey"
    
    message("Plot 1: UMAP")
    plot1 = ggplot(data.frame(so@meta.data, dim1 = dim1, dim2 = dim2)) + 
        geom_point_rast(aes(dim1, dim2, color = color_by), alpha = I(alpha), raster.dpi = 100) +
        theme_bw() + scale_color_manual(values = mypal)
    print(plot1)
    
    message("Plot 2: UMAP split")
    tmp = data.frame(so@meta.data, dim1 = dim1, dim2 = dim2)
    bkgrd = data.frame(dim1 = dim1, dim2 = dim2)
    
    p = ggplot(bkgrd) + geom_point_rast(aes(dim1, dim2), color = "grey", size = 0.1, alpha = 0.2, raster.dpi = 50)
    p2 = geom_point_rast(data = tmp, aes(dim1, dim2, color = color_by), size = 1,  alpha = alpha) 
    
    if (is.null(split_by2)) {
        plot2 = p + p2 + scale_color_manual(values = mypal) + theme_bw() + facet_wrap(facets = vars(so@meta.data[,"split_by1"])) + ggtitle(label = sprintf("color: %s, grid: %s", color_by, split_by1))
    } else {
        plot2 = p + p2 + scale_color_manual(values = mypal)  + theme_bw() + facet_grid(rows = vars(so@meta.data[,"split_by1"]), cols = vars(so@meta.data[,"split_by2"])) + ggtitle(label = sprintf("color: %s, grid: %s x %s", color_by, split_by1, split_by2))
        }
    print(plot2)
    
    message("Plot 3: UMAP genes")
    
    for (g in genes) {
        print(g)
        tmp = data.frame(so@meta.data, dim1 = dim1, dim2 = dim2, size = so@assays$RNA$counts[rownames(so@assays$RNA$counts) %in% g,])
        bkgrd = data.frame(dim1 = dim1, dim2 = dim2)
        
        p = ggplot(bkgrd) + geom_point_rast(aes(dim1, dim2), color = "grey", size = 0.1, raster.dpi = 50)
        p2 = geom_point(data = tmp, aes(dim1, dim2, color = size > 0, size = size,  alpha = size > 0)) 
        #p2 =  geom_point(data = tmp2, aes(dim1, dim2, color = size > 0, size = size), alpha = I(alpha))  + scale_color_manual
    
        
        if (is.null(split_by2)) {
        plot3 = p + p2  + 
            scale_color_manual(values = c("black", "red")) + 
            scale_alpha_manual(values = c(0.5,1))  + 
            theme_bw()  + 
            facet_wrap(facets = vars(so@meta.data[,"split_by1"])) + 
            ggtitle(label = sprintf("gene: %s, color: %s, grid: %s", g, color_by, split_by1))
    } else {
         plot3 = p + p2  + 
             scale_color_manual(values = c("black", "red")) + 
             scale_alpha_manual(values = c(0.5,1))  + 
             theme_bw() + 
             facet_grid(rows = vars(so@meta.data[,"split_by1"]), cols = vars(so@meta.data[,"split_by2"])) +
             ggtitle(label = sprintf("gene: %s, color: %s, grid: %s", g, color_by, split_by1))
        }
        print(plot3)
        
    }

}

ConvertS5toS3 = function(so, assay1 = "RNA", assay2 = "ADT") {
    so.v3 = CreateAssayObject(counts = so[[assay1]]$counts)
    so.v3 = CreateSeuratObject(so.v3)
    so.v3[[assay2]] = CreateAssayObject(counts = so[[assay2]]$counts) #samples@assays$ADT$counts
    #print(all(colnames(so.v3@assays$RNA$counts) == colnames(so.v3@assays$ADT$counts)))
    #all(colnames(so.v3) == colnames(so))
    so.v3@meta.data = so@meta.data
    return(so.v3)
}

Vplot = function(vplot, xlab = "FC", xlimits = c(0.1, 10), ylimits = c(10^-300,1)) {
  p = ggplot(data = vplot) + geom_point_rast(aes(x = fc, y = pval), colour = "grey", alpha = I(1), size = I(1), raster.dpi = 100) +
  scale_x_continuous(trans = log_trans(10), limits = xlimits) + #breaks = c(1/50,1/40, 1/30, 1/20, 1/10, 1/5,1/2,1,2,5,10, 20, 30, 40, 50),labels =c("1/50","1/40", "1/30", "1/20", "1/10", "1/5","1/2","1","2","5","10", "20", "30", "40", "50")
  scale_y_continuous(trans = log_trans(10), breaks = trans_breaks("log10", function(x) 10^x), labels = trans_format("log10", math_format(10^.x)), limits = ylimits) + #limits = c(10^-5.5, 1)
  #geom_hline(aes(yintercept = 0.05), linetype="dashed", color = "brown") +
  annotation_logticks(sides = "b") +
  xlab(xlab) +
  theme_bw() +
  ylab("p value") +
  theme(axis.text.x  = element_text(size=20,angle = 0, hjust = 0.5), axis.text.y  = element_text(size=20), legend.text=element_text(size=20), axis.title.x = element_text(size=20) , axis.title.y = element_text(size=20))
  return(p)
}

FCFCplot = function(vplot, xlab = "fc1", ylab = "fc2", main = "", printgeomtext = T, xlimits = c(0.1, 10),  ylimits = c(0.1, 10)) {
  p = ggplot(data = vplot) + geom_point(aes(x = fc1, y = fc2), colour = "black", alpha = I(1), size = I(0.5)) +
    scale_x_continuous(trans = log_trans(10), breaks = c(0.125,0.5, 1, 2,5, 10),  labels = c(0.125,0.5, 1, 2,5, 10), limits = xlimits) +
    scale_y_continuous(trans = log_trans(10), breaks = c(0.125,0.5, 1, 2,5, 10),  labels = c(0.125,0.5, 1, 2,5, 10), limits = ylimits) +
    geom_hline(aes(yintercept = 1), linetype="dashed", color = "brown") +
    geom_vline(aes(xintercept = 1), linetype="dashed", color = "brown") +
  #geom_abline(intercept = 0, slope = 1, linetype="dashed", color = "brown") +
    annotation_logticks(sides = "bl") +
    xlab(xlab) +
    ylab(ylab) +
    ggtitle(main) +
    theme_bw() +
    theme(axis.text.x  = element_text(size=15,angle = 0, hjust = 1), axis.text.y  = element_text(size=15), legend.text=element_text(size=20), axis.title.x = element_text(size=20) , axis.title.y = element_text(size=20))
  return(p)
}

SplitColors = function(pal = mypal[1:length(levels(cds$clustersCCA9))], splitvector = sapply(levels(cds$clustersCCA9), function(x) { length(which(grepl(x, levels(cds$clustersCCA9Split))))}) ) {
  newpal = c()
  for (c in 1:length(splitvector)){
    #print(c)
    n = splitvector[c]
    newcol = matrix(rep(col2rgb(pal[c], alpha = T), n), ncol = n)
    newcol[4,] = seq(50, 255, length.out = n)
    newpal = c(newpal, rgb(red = newcol[1,]/255, blue = newcol[2,]/255, green = newcol[3,]/255, alpha = newcol[4,]/255))
  }
  return(newpal)
}
sp = SplitColors
#pie(1:length(levels(cds$clustersCCA9Split)), col = sp)

library(org.Hs.eg.db)
hs = org.Hs.eg.db
go2eg = as.list(org.Hs.egGO2ALLEGS)
symbols = select(hs, keys = unique(unlist(go2eg)), columns = "SYMBOL", keytype = "ENTREZID")
cc = sort(symbols$SYMBOL[symbols$ENTREZID %in% unique(c(go2eg[["GO:0007049"]]))])
cellmembrane = sort(symbols$SYMBOL[symbols$ENTREZID %in% unique(unlist(list(go2eg[["GO:0009986"]], go2eg[["GO:0009897"]], go2eg[["GO:0016494"]], go2eg[["GO:0098552"]])))]) #go2eg[["GO:0016020"]]
#is_cellmembrane = rownames(cor_genes) %in% is_cellmembrane
tf =  sort(symbols$SYMBOL[symbols$ENTREZID %in% unique(go2eg[["GO:0003700"]])])
secretedmol = sort(symbols$SYMBOL[symbols$ENTREZID %in% unique(go2eg[["GO:0005615"]])])
effmol = sort(symbols$SYMBOL[symbols$ENTREZID %in% unique(go2eg[["GO:0005615"]])])
chemreceptor = symbols$SYMBOL[symbols$ENTREZID %in% unique(c(go2eg[["GO:0004950"]]))]

library(org.Mm.eg.db)
mm = org.Mm.eg.db
go2eg = as.list(org.Mm.egGO2ALLEGS)
symbols = select(mm, keys = unique(unlist(go2eg)), columns = "SYMBOL", keytype = "ENTREZID")
cc = sort(symbols$SYMBOL[symbols$ENTREZID %in% unique(c(go2eg[["GO:0007049"]]))])
cellmembrane = sort(symbols$SYMBOL[symbols$ENTREZID %in% unique(unlist(list(go2eg[["GO:0009986"]], go2eg[["GO:0009897"]], go2eg[["GO:0016494"]], go2eg[["GO:0098552"]])))]) #go2eg[["GO:0016020"]]
#is_cellmembrane = rownames(cor_genes) %in% is_cellmembrane
tf =  sort(symbols$SYMBOL[symbols$ENTREZID %in% unique(go2eg[["GO:0003700"]])])
effmol = sort(symbols$SYMBOL[symbols$ENTREZID %in% unique(go2eg[["GO:0005615"]])])
chemreceptor = symbols$SYMBOL[symbols$ENTREZID %in% unique(c(go2eg[["GO:0004950"]]))]

```

**Treg data subsetting**
See integration_midway3.Rmd for Treg data subsetting and TOTALVI analysis
Here for reference only:
```{r}
#Tregs
setwd("/project/jfkfloor2/zemmourlab/david/immgent/analysis/integration/IGT1_56")
SubsetData2 = function(so = so, to_subset = "resting_CD4") {
    tmp = so[,so$ClusterTOTALVI_Res1 %in% as.character(annot$ClusterTOTALVI_Res1[annot$annotation_broad == to_subset])]
    tmp[["RNA"]]$counts = as(tmp[["RNA"]]$counts, Class = "dgCMatrix")
    tmp[["ADT"]]$counts = as(tmp[["ADT"]]$counts, Class = "dgCMatrix")
    tmp[["umap_totalvi_20231030"]] = tmp[["umap_totalvi"]]
    tmp[["totalvi_20231030"]] = tmp[["totalvi"]]
    message(sprintf("Dimensions of subset: %s cells", ncol(tmp)))
    dir.create(to_subset)
    message("Saving Rds")
    saveRDS(tmp, file = sprintf("%s/immgent_IGT1_56_%s.Rds", to_subset, to_subset))
}
to_subset = "Treg"
SubsetData2(so, to_subset)
#Run scripts: create the wrapper and modify the lines as follows:
prefix = sprintf("totalvi_igt1_56_allgenes_%s_20240306", to_subset)
#SCRIPT_DIR=/project/jfkfloor2/zemmourlab/david/immgent/immgent_integration_git
sprintf("sbatch $SCRIPT_DIR/run_totalvi_wrapper_20240306_%s.sh", to_subset) #run the 2 following scripts: 
sprintf("Rscript run_totalvi.R /project/jfkfloor2/zemmourlab/david/immgent/analysis/integration/IGT1_56/%s/ immgent_IGT1_56_%s.Rds totalvi_igt1_56_allgenes_%s_20240306", to_subset, to_subset, to_subset)
sprintf("Rscript run_totalvi_plot.R /project/jfkfloor2/zemmourlab/david/immgent/analysis/integration/IGT1_56/%s/ totalvi_igt1_56_allgenes_%s_20240306", to_subset, to_subset)

```

**Load data in R**
```{r}
#IGT1-56 subsets
prefix = "totalvi_igt1_56_20240306_allgenes"
to_subset = "Treg"
so = readRDS(file = sprintf("/project/jfkfloor2/zemmourlab/david/immgent/analysis/integration/IGT1_56/%s/bpcells/totalvi_igt1_56_allgenes_%s_20240306_counts/totalvi_igt1_56_allgenes_%s_20240306_BPCells_so.Rds", to_subset, to_subset, to_subset))
# so = readRDS(file = sprintf("~/google_drive/ImmgenT/analysis/data_integration/IGT1_56/%s/bpcells/totalvi_igt1_56_allgenes_%s_20240306_counts/totalvi_igt1_56_allgenes_%s_20240306_BPCells_so.Rds", to_subset, to_subset, to_subset))
```

Plots
```{r}
prefix = paste(prefix, to_subset, sep = "_")
sample_qc = read.table("~/google_drive/ImmgenT/analysis/QC/Sample_metadata_withQC.csv", header = T, sep = ",")
sample_qc$IGT = sample_qc$immgent_IGT_10xlane_id
sample_qc$sample_id = paste(sample_qc$immgent_IGT_10xlane_id, sample_qc$hashtag_number, sep = "_")

so$sample_id = paste(so$IGT, so$hashtag_number, sep = "_")

#sample_qc$organ[match(so$sample_id, sample_qc$sample_id)]
so$organ_clean = sample_qc$organ[match(so$sample_id, sample_qc$sample_id)]
so$organ_clean[is.na(so$organ_clean)] = so$organ[is.na(so$organ_clean)]
#table(so$organ[which(is.na(so$organ_clean))])
table(so$organ_clean, so$organ)
so$organ_clean_simplified = so$organ_clean
so$organ_clean_simplified[grepl(pattern = "lymph node", x = so$organ_clean)] = "lymph node"
so$organ_clean_simplified[grepl(pattern = "duodenum|ileum|jejunum|Lamina|SI|small intestine", x = so$organ_clean)] = "small intestine"
so$organ_clean_simplified[grepl(pattern = "gut", x = so$organ_clean)] = "colon"
so$organ_clean_simplified[grepl(pattern = "salivary", x = so$organ_clean)] = "submandibular gland"
so$organ_clean_simplified[grepl(pattern = "skin", x = so$organ_clean)] = "skin"
table(so$organ_clean_simplified)

so$condition = sample_qc$condition[match(so$sample_id, sample_qc$sample_id)]
so$condition[is.na(so$condition) & grepl(pattern = "LCMV",x = so$sample_name.1)] = "virus"
so$condition[is.na(so$condition)] = "healthy"
table(so$condition)

pdf(sprintf("%s_UMAP_organ_condition.pdf", prefix), 12,10, useDingbats = F)
DimPlot(so, reduction = "umap_totalvi", group.by = "IGT", raster = F) + scale_color_manual(values = mypal)
DimPlot(so, reduction = "umap_totalvi", group.by = "organ_clean_simplified", raster = F) + scale_color_manual(values = mypal) + scale_color_manual(values = mypal)
DimPlot(so, reduction = "umap_totalvi", group.by = "condition", raster = F) + scale_color_manual(values = mypal) + scale_color_manual(values = mypal)

DimPlot(so, reduction = "umap_totalvi", split.by = "organ_clean_simplified", group.by = "IGT", ncol = 5, raster = F)  + scale_color_manual(values = mypal)
DimPlot(so, reduction = "umap_totalvi", split.by = "organ_clean_simplified", group.by = "condition", ncol = 5, raster = F)  + scale_color_manual(values = mypal)

DimPlot(so, reduction = "umap_totalvi", split.by = "condition", group.by = "IGT", ncol = 3, raster = F) + scale_color_manual(values = mypal)
DimPlot(so, reduction = "umap_totalvi", split.by = "condition", group.by = "organ_clean_simplified", ncol = 3, raster = F) + scale_color_manual(values = mypal)
dev.off()

so = NormalizeData(so,assay = "ADT", normalization.method = "CLR", verbose = T)
so = NormalizeData(so,assay = "RNA", normalization.method = "LogNormalize",verbose = T)

pdf(sprintf("%s_UMAP_genes.pdf", prefix), 12,10, useDingbats = F)
rstr = F
FeaturePlot(so, slot = "data", features = "Foxp3", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, slot = "data", features = "Il2ra", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, features = "IL2RA.CD25", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, slot = "data", features = "Ctla4", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, slot = "data", features = "Tnfrsf18", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, slot = "data", features = "GITR.CD357", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, slot = "data", features = "Izumo1r", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, slot = "data", features = "FR4", raster = rstr, pt.size = 2, order = T)

FeaturePlot(so, features = "Ccr7", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, features = "Sell", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, features = "CD62L", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, features = "CD44", raster = rstr, pt.size = 2, order = T)

FeaturePlot(so, features = "NEUROPILIN1.CD304", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, features = "Rorc", raster = rstr, pt.size = 2, order = T)

FeaturePlot(so, features = "Areg", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, features = "Il1rl1", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, features = "Pparg", raster = rstr, pt.size = 2, order = T)

FeaturePlot(so, features = "Mki67", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, features = "Tbx21", raster = rstr, pt.size = 2, order = T)
FeaturePlot(so, features = "Nr4a1", raster = rstr, pt.size = 2, order = T)
dev.off()

```

**Method 1: Hotspot**
based on latent space proximity
caveat: each gene can be at most in one module, in contrasts to topic modeling
Installation: see install_scvi.Rmd

```{python}
import warnings; warnings.simplefilter('ignore')

import scvi
import os
import hotspot
import scanpy as sc
import muon as mu

import numpy as np
import mplscience
import matplotlib.pyplot as plt
import pickle

os.chdir("/project/jfkfloor2/zemmourlab/david/immgent/analysis/integration/IGT1_56/Treg")

n_var_genes = 2000

#adata = sc.read_h5ad("totalvi_igt1_56_allgenes_Treg_20240306/adata.h5ad")
model = scvi.model.TOTALVI.load("totalvi_igt1_56_allgenes_Treg_20240306/")
mdata = mu.MuData({"RNA": model.adata})
mdata.mod["RNA"].layers["counts"] = mdata.mod["RNA"].X.copy()
protein_data = model.adata.obsm['protein']
protein_adata = sc.AnnData(X=protein_data)
mdata.mod['protein'] = protein_adata
del mdata["RNA"].obsm['protein']
TOTALVI_LATENT_KEY = "X_totalVI"
mdata.mod["RNA"].obsm[TOTALVI_LATENT_KEY] = model.get_latent_representation()



adata = mdata.mod["RNA"]
#Remove genes with no counts
gene_sums = np.sum(adata.X, axis=0)
genes_non_zero = gene_sums > 0
adata = adata[:, genes_non_zero]
adata.layers["counts"] = adata.X.copy()
adata.layers["counts_csc"] = adata.layers["counts"].tocsc()

hs = hotspot.Hotspot(
    adata,
    layer_key="counts_csc",
    model='danb',
    latent_obsm_key="X_totalVI"
)

hs.create_knn_graph( weighted_graph=False, n_neighbors=30)

#Find most variable genes by autocorrelation
hs_results = hs.compute_autocorrelations(jobs=16)
hs_results.head(15)
hs.results.to_csv('autocorrelations.csv', sep=',', header=True, index=True)

hs_genes = hs_results.loc[hs_results.FDR < 0.05].sort_values('Z', ascending=False).head(2000).index

#Compute local correlation in the TOTALVI latent space
lcz = hs.compute_local_correlations(hs_genes, jobs=16)
lcz.to_csv('local_correlations_2000genes.csv', sep=',', header=True, index=True)

#Find modules
modules = hs.create_modules(
    min_gene_threshold=15, core_only=True, fdr_threshold=0.05
)

plt.figure(figsize=(15,15 ))
hs.plot_local_correlations(vmin=-12, vmax=12)
plt.savefig("modules_2000genes_heatmap.pdf", format='pdf', bbox_inches='tight')
plt.close() 

results = hs.results.join(hs.modules) #merges both pd df
results.to_csv('modules_2000genes_table.csv', sep=',', header=True, index=True)

# Show the top genes for a module
module = 1
results2 = results.loc[results.Module == module]
results2.sort_values('Z', ascending=False).head(50)

#Calculate module score for each single cell
module_scores = hs.calculate_module_scores()
module_scores.head()
module_scores.to_csv('modules_2000genes_singlecellscores.csv', sep=',', header=True, index=True)

sc.pp.neighbors(adata, use_rep = "X_totalVI")
sc.tl.umap(adata)
#sc.pl.umap(adata, frameon=False)

module_cols = []
for c in module_scores.columns:
    key = f"Module {c}"
    adata.obs[key] = module_scores[c]
    module_cols.append(key)
    
with mplscience.style_context():
    sc.pl.umap(adata, color=module_cols, frameon=False, vmin=-1, vmax=1, save='UMAP_module_plot.pdf')
    


#save object
# Serialize and save the `hs` object to disk
with open('hs_object.pkl', 'wb') as file:
    pickle.dump(hs, file)
# Later, you can load the `hs` object back into memory
with open('hs_object.pkl', 'rb') as file:
    hs = pickle.load(file)

```

Plot the modules in the UMAP
```{r}
setwd("~/google_drive/ImmgenT/analysis/data_integration/IGT1_56/Treg")
module_scores = read.csv("~/google_drive/ImmgenT/analysis/data_integration/IGT1_56/Treg/modules_2000genes_singlecellscores.csv", row.names = 1, header = T, sep = ",")

module_genes = read.csv("~/google_drive/ImmgenT/analysis/data_integration/IGT1_56/Treg/modules_2000genes_extended.csv", header = T, sep = ",")
module_genes$Module = sprintf("Module_%s",module_genes$Module)

colnames(module_scores) = gsub(pattern = "X", replacement = "Module_", colnames(module_scores))
module_scores = module_scores[match(rownames(so@meta.data), rownames(module_scores)),]
all(rownames(module_scores) == rownames(so@meta.data))
so@meta.data = merge(so@meta.data, module_scores, by = "row.names", all.x = TRUE)
rownames(so@meta.data) = so@meta.data[,"Row.names"]


#so[["m1"]]
#tmp = CreateAssay5Object(data = Matrix::Matrix(t(module_scores), sparse = TRUE))
#tmp@meta.data = so@meta.data
#rownames(tmp@meta.data) = tmp@meta.data$Row.names
#tmp[["umap_totalvi"]] = so[["umap_totalvi"]]
all(rownames(so@meta.data) == rownames(so[["umap_totalvi"]]))
all(rownames(so@meta.data) == colnames(so))

#module = "Module_1"
pdf("modules_UMAP.pdf", 9,8, useDingbats = F)
for (module in colnames(module_scores) ) {
    print(module)
    genes = module_genes %>% filter(Module == module) %>% pull(Gene)
    tmp = data.frame(so@meta.data, dim1 = so[["umap_totalvi"]]@cell.embeddings[match(colnames(so),rownames(so[["umap_totalvi"]])),1], dim2 = so[["umap_totalvi"]]@cell.embeddings[match(colnames(so),rownames(so[["umap_totalvi"]])),2], size = so@meta.data[,module]) 
    tmp$size[tmp$size > 1] = 1
    tmp$size[tmp$size < -1] = -1
    
    p = ggplot(tmp) + geom_point(aes(dim1, dim2, color = size), size = 2, alpha = I(alpha)) + theme_bw() + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + scale_color_gradientn(colours = ColorRamp)+ theme(axis.text.x  = element_text(size=15), axis.text.y  = element_text(size=15), legend.text=element_text(size=15), axis.title.x = element_text(size=20) , axis.title.y = element_text(size=20), legend.title = element_blank()) + ggtitle(label = sprintf("%s:%s...",module, paste(genes[1:10], collapse=",")))
    print(p)
}
dev.off()

```
Plot modules in single cells
```{r}
setwd("~/google_drive/ImmgenT/analysis/data_integration/IGT1_56/Treg")

module_scores = read.csv("~/google_drive/ImmgenT/analysis/data_integration/IGT1_56/Treg/modules_2000genes_singlecellscores.csv", row.names = 1, header = T, sep = ",")
colnames(module_scores) = gsub(pattern = "X", replacement = "Module_", colnames(module_scores))
module_scores = module_scores[match(rownames(so@meta.data), rownames(module_scores)),]
all(rownames(module_scores) == rownames(so@meta.data))
so@meta.data = merge(so@meta.data, module_scores, by = "row.names", all.x = TRUE)
rownames(so@meta.data) = so@meta.data[,"Row.names"]

percentiles = function(vector = c(1) ) {return(ecdf(vector)(vector) * 100) }
module_scores_percentile = apply(module_scores, 2, percentiles)
rownames(module_scores_percentile) = rownames(module_scores)
module_scores_percentile2 = module_scores_percentile / rowSums(module_scores_percentile) *100


so = RunUMAP(so, dims = 1:30, reduction = "totalvi", n.components = 1, reduction.name = "totalvi_1d")
ord_cells = order(so[["totalvi_1d"]]@cell.embeddings[,1])

hc = hclust(dist(t(module_scores_percentile2)))
ord_modules = colnames(module_scores_percentile2)[hc$order]

#pca = prcomp(module_scores_percentile2, scale. = TRUE)
#pca_1 = pca$x[, 1]
#ordered_df = as.data.frame(module_scores_percentile2[order(pca_1),])
ordered_df = as.data.frame(module_scores_percentile2[ord_cells,ord_modules])
ordered_df$cell_id = factor(rownames(ordered_df), ordered = T)
melted_df = melt(ordered_df)
melted_df$variable = factor(melted_df$variable, levels = ord_modules, ordered = T)
melted_df = melted_df %>% arrange(cell_id)

pdf("modules_singlecell_barplot.pdf", 20,10, useDingbats = F)
p = ggplot(melted_df, aes(x = cell_id, y = value, fill = variable)) +
    geom_bar(stat = "identity") +
    scale_fill_manual(values = mypal)+
    labs(x = "Cells", y = "Score", fill = "Module") +
    theme_minimal() +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_blank()
    )
print(p)
dev.off()


```


